packages/hub/src/utils/symlink.ts

```diff
@@ -0,0 +1,172 @@
+	if (path.startsWith("~")) {
+		return path.replace("~", os.homedir());
+	}
+	return path;
+}
+
+function commonPath(...paths: [string, ...string[]]): string {
+	// Normalize paths
+	paths = paths.map((p) => path.resolve(p)) as [string, ...string[]];
+
+	const parts = paths.map((p) => p.split(path.sep));
+	const common = parts[0];
+
+	for (let i = 1; i < parts.length; i++) {
+		for (let j = 0; j < common.length; j++) {
+			if (parts[i][j] !== common[j]) {
+				common.length = j;
+				break;
+			}
+		}
+	}
+
+	return common.join(path.sep) || path.sep;
+}
+
+/**
+ * Return whether the symlinks are supported on the machine.
+ *
+ * Since symlinks support can change depending on the mounted disk, we need to check
+ * on the precise cache folder. By default, the default HF cache directory is checked.
+ */
+async function areSymlinksSupported(cache_dir: string | undefined): Promise<boolean> {
+	if (!cache_dir) {
+		cache_dir = getHFHubCachePath();
+	}
+	cache_dir = path.resolve(expandUser(cache_dir));
+	if (!_are_symlinks_supported_in_dir.has(cache_dir)) {
+		_are_symlinks_supported_in_dir.set(cache_dir, true);
+
+		await fs.mkdir(cache_dir, { recursive: true });
+		const tmp_dir = await fs.mkdtemp(cache_dir);
+		try {
+			const src_path = path.join(tmp_dir, "dummy_src");
+			await fs.open(src_path, "w").then((fd) => fd.close());
+			const dst_path = path.join(tmp_dir, "dummy_dst");
+
+			const relative_src = path.relative(src_path, path.dirname(dst_path));
+			try {
+				await fs.symlink(dst_path, relative_src);
+			} catch (_e: unknown) {
+				_are_symlinks_supported_in_dir.set(cache_dir, false);
+
+				if (!process.env.HF_HUB_DISABLE_SYMLINKS_WARNING) {
+					let message = `
+					\`huggingface_hub\` cache-system uses symlinks by default to
+					efficiently store duplicated files but your machine does not
+					support them in ${cache_dir}. Caching files will still work
+					but in a degraded version that might require more space on
+					your disk. This warning can be disabled by setting the
+					\`HF_HUB_DISABLE_SYMLINKS_WARNING\` environment variable. For
+					more details, see
+					https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
+					`;
+					if (os.platform() === "win32") {
+						message += `
+						\nTo support symlinks on Windows, you either need to
+						activate Developer Mode or to run Python as an
+						administrator. In order to activate developer mode,
+						see this article:
+						https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
+						`;
+					}
+					console.warn(message);
+				}
+			}
+		} finally {
+			await fs.rm(tmp_dir, { recursive: true });
+
+			if (_are_symlinks_supported_in_dir.size > MAX_CACHE_SIZE) {
+				const key = _are_symlinks_supported_in_dir.keys().next().value;
+				if (key) {
+					_are_symlinks_supported_in_dir.delete(key);
+				}
+			}
+		}
+	}
+	return _are_symlinks_supported_in_dir.get(cache_dir) ?? false;
+}
```

Comment:
Is it possible to remove this whole function?

By doing in `createSymlink` something like:

```ts
	try {
    await symlink(dest, src);
    return;
  } catch {}

  // symlink failed, copy or rename the file
```